### PR TITLE
curvefs: change the data organization format

### PR DESCRIPTION
**Pull request by 201341**
Signed-off-by: swj <1186093704@qq.com>

### What problem does this PR solve?
To improve the s3 data storage sync performance

Issue Number: #2380 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by 201341** [comment]
cicheck

**Comment by wuhongsong** [comment]
image

@201341  cpplint is failed, you can   run cpplint locally like `cpplint hello.cpp`

File: curvefs/src/common/s3util.h
Comment (wuhongsong) [inline curvefs/src/common/s3util.h:38]: fsid repeated？

**Comment by 201341** [comment]
> image
> 
> @201341 cpplint is failed, you can run cpplint locally like `cpplint hello.cpp`

ok, I will fix it later

File: curvefs/src/common/s3util.h
Comment (wuhongsong) [inline curvefs/src/common/s3util.h:35]: Two suggestions:
1. Maybe we don't need the directory prefix as the object name？
2. maybe we can use hash or % replace ？

File: curvefs/src/client/s3/disk_cache_read.cpp
Comment (wuhongsong) [inline curvefs/src/client/s3/disk_cache_read.cpp:99]: can we move the mkdir to DiskCacheBase?

File: curvefs/src/common/s3util.h
Comment (201341) [reply curvefs/src/common/s3util.h:35]: Hi, Adding directory prefix has three advantages.
1:  improve the s3 data storage sync performance when we sync different cloud provider using the tools rclone
2: improve s3 write performance when deploying on aws as the [doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) suggested
3: improve disk cache write performance.
So I suggerest using the directory prefix.

Comment (wuhongsong) [reply curvefs/src/common/s3util.h:35]: > Hi, Adding directory prefix has three advantages. 1: improve the s3 data storage sync performance when we sync different cloud provider using the tools rclone 2: improve s3 write performance when deploying on aws as the [doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) suggested 3: improve disk cache write performance as writing 1000w file into 1 directory maybe has poor write performance. So I suggerest using the directory prefix.


yeah, you are right. lets do it.  but can we use hash or % replace `inodeid/1000` ? then the files will more evenly distributed,？

Comment (201341) [reply curvefs/src/common/s3util.h:35]: > > Hi, Adding directory prefix has three advantages. 1: improve the s3 data storage sync performance when we sync different cloud provider using the tools rclone 2: improve s3 write performance when deploying on aws as the [doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) suggested 3: improve disk cache write performance as writing 1000w file into 1 directory maybe has poor write performance. So I suggerest using the directory prefix.
> 
> yeah, you are right. lets do it. but can we use hash or % replace `inodeid/1000` ? then the files will more evenly distributed,？

yep, maybe we can make it an optional as [juicefs](https://github.com/juicedata/juicefs/blob/main/pkg/chunk/cached_store.go#L74) does.

Comment (wuhongsong) [reply curvefs/src/common/s3util.h:35]: good job, lets do it.

Comment (wuhongsong) [reply curvefs/src/common/s3util.h:35]: > > > Hi, Adding directory prefix has three advantages. 1: improve the s3 data storage sync performance when we sync different cloud provider using the tools rclone 2: improve s3 write performance when deploying on aws as the [doc](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) suggested 3: improve disk cache write performance as writing 1000w file into 1 directory maybe has poor write performance. So I suggerest using the directory prefix.
> > 
> > 
> > yeah, you are right. lets do it. but can we use hash or % replace `inodeid/1000` ? then the files will more evenly distributed,？
> 
> yep, maybe we can make it an optional as [juicefs](https://github.com/juicedata/juicefs/blob/main/pkg/chunk/cached_store.go#L74) does.

good job, lets do it

File: curvefs/src/common/s3util.h
Comment (ilixiaocui) [inline curvefs/src/common/s3util.h:36]: Compatibility with existing systems needs to be considered.

File: curvefs/src/common/s3util.h
Comment (ilixiaocui) [inline curvefs/src/common/s3util.h:37]: We need a comment that exemplifies what each field means.

Comment (ilixiaocui) [reply curvefs/src/common/s3util.h:37]: Why do you need inodeid/1000 and inodeid/1000/1000?

Comment (wuhongsong) [reply curvefs/src/common/s3util.h:37]: > Why do you need inodeid/1000 and inodeid/1000/1000?
for  multi directory and  s3 prefix，but its better to  replace it with hash or %

File: curvefs/src/common/s3util.h
Comment (201341) [reply curvefs/src/common/s3util.h:36]: There is no better way to handle with compatibility. If changing the data organization, we have to sync the existing file system to new file system.

**Comment by ilixiaocui** [comment]
@201341 What is the current status of this PR?Whether the review opinions have been modified or have any questions？

**Comment by 201341** [comment]
@ilixiaocui I'm working on this review opinions. For compatibility with existing systems, I have no good idea, so I don't push this code.

**Comment by ilixiaocui** [comment]
> @ilixiaocui I'm working on this review opinions. For compatibility with existing systems, I have no good idea, so I don't push this code.

For deployed clusters, it should not be realistic to format all existing objects.

One idea is it supports both encodings, which brings a problem. Every time you go to s3 to get it, you have to get it twice according to different shard names.

The second method is to add a flag configuration, the old cluster has always used the old encoding method, and the newly deployed cluster opens the new configuration.In this way, the old cluster cannot be upgraded to the new encoding method.

**Comment by 201341** [comment]
> > @ilixiaocui I'm working on this review opinions. For compatibility with existing systems, I have no good idea, so I don't push this code.
> 
> For deployed clusters, it should not be realistic to format all existing objects.
> 
> One idea is it supports both encodings, which brings a problem. Every time you go to s3 to get it, you have to get it twice according to different shard names.
> 
> The second method is to add a flag configuration, the old cluster has always used the old encoding method, and the newly deployed cluster opens the new configuration.In this way, the old cluster cannot be upgraded to the new encoding method.

Second method sounds good, I'll do it.

**Comment by ilixiaocui** [comment]
> > > @ilixiaocui I'm working on this review opinions. For compatibility with existing systems, I have no good idea, so I don't push this code.
> > 
> > 
> > For deployed clusters, it should not be realistic to format all existing objects.
> > One idea is it supports both encodings, which brings a problem. Every time you go to s3 to get it, you have to get it twice according to different shard names.
> > The second method is to add a flag configuration, the old cluster has always used the old encoding method, and the newly deployed cluster opens the new configuration.In this way, the old cluster cannot be upgraded to the new encoding method.
> 
> Second method sounds good, I'll do it.

The second method has an update, and the updated one should be more reasonable.

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
pjdtest failed, As I change the client.conf, the curveadm can't mount the filesystem successfully, I can't handle this

**Comment by 201341** [comment]
@wuhongsong @ilixiaocui  would you mind spend some time on reviewing this pr?

**Comment by ilixiaocui** [comment]
> @wuhongsong @ilixiaocui would you mind spend some time on reviewing this pr?

Sorry for making you wait so long. We will review as soon as possible. I will add this issue to the developer event in 2023Q2.

**Comment by wuhongsong** [comment]
sorry, I will review it next week, so busy lately.

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/tools/curvefs_tool_define.cpp
Comment (wuhongsong) [inline curvefs/src/tools/curvefs_tool_define.cpp:256]: s3.objectPrefix needed

**Comment by wuhongsong** [comment]
> pjdtest failed, As I change the client.conf, the curveadm can't mount the filesystem successfully, I can't handle this

I have looked at the reasons for the pjdtest failure. there is one error in s3.objectprefix, s3.objectPrefix needed.

**Comment by 201341** [comment]
> > pjdtest failed, As I change the client.conf, the curveadm can't mount the filesystem successfully, I can't handle this
> 
> I have looked at the reasons for the pjdtest failure. there is one error in s3.objectprefix, s3.objectPrefix needed.

yes, s3.objectPrefix is needed, and I have to modify the curveadm's code to support this after this pr is merged just as github can't merge the two repo's pr at the same time.

**Comment by wuhongsong** [comment]
> > pjdtest failed, As I change the client.conf, the curveadm can't mount the filesystem successfully, I can't handle this
> 
> I have looked at the reasons for the pjdtest failure. there is one error in s3.objectprefix, s3.objectPrefix needed.

root@pubbeta2-curve3:/data/logs/curvefs# sudo docker logs 27c3b00eb625
F 2023-04-07T11:34:25.596459+0800 196976 configuration.cpp:318] Get s3.objectprefix from /etc/curvefs/tools.conf fail
*** Check failure stack trace: ***
*** Aborted at 1680838465 (unix time) try "date -d @1680838465" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x30170) received by PID 196976 (TID 0x7f2102de7240) from PID 196976; stack trace: ***
    @     0x7f21010a00e0 (unknown)
    @     0x7f20ff757fff gsignal
    @     0x7f20ff75942a abort
    @     0x55fcf3844f54 google::FlushAndAbort()
    @     0x55fcf38421a6 google::LogMessage::Fail()
    @     0x55fcf38420ef google::LogMessage::SendToLog()
    @     0x55fcf3841918 google::LogMessage::Flush()
    @     0x55fcf38448e8 google::LogMessageFatal::~LogMessageFatal()
    @     0x55fcf333f9ec curve::common::Configuration::GetValueFatalIfFail()
    @     0x55fcf305fa1b curvefs::tools::SetDiffFlagInfo<>()
    @     0x55fcf3068272 _ZNSt5_BindIFPFvPN5curve6common13ConfigurationEPN6gflags19CommandLineFlagInfoERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESE_PjESt12_PlaceholderILi1EESI_ILi2EEPKcSM_SF_EE6__callIvJOS3_OS6_EJLm0ELm1ELm2ELm3ELm4EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE
    @     0x55fcf3066b14 std::_Bind<>::operator()<>()
    @     0x55fcf3064c60 std::_Function_handler<>::_M_invoke()
    @     0x55fcf30a3141 std::function<>::operator()()
    @     0x55fcf30a1224 curvefs::tools::CurvefsToolRpc<>::UpdateFlags()
    @     0x55fcf30a072d curvefs::tools::CurvefsToolRpc<>::Init()
    @     0x55fcf315215d curvefs::tools::create::CreateFsTool::Init()
    @     0x55fcf3144b99 curvefs::tools::CurvefsTool::Run()
    @     0x55fcf3057b39 main
    @     0x7f20ff7452e1 __libc_start_main
    @     0x55fcf30577aa _start
    @                0x0 (unknown)
/client.sh: line 9: 196976 Aborted                 (core dumped) $g_curvefs_tool $g_curvefs_tool_operator $g_fsname $g_fstype $g_rpc_timeout_ms
CREATEFS FAILED
root@pubbeta2-curve3:/data/logs/curvefs#

**Comment by Cyber-SiKu** [comment]
> > > pjdtest failed, As I change the client.conf, the curveadm can't mount the filesystem successfully, I can't handle this
> > 
> > 
> > I have looked at the reasons for the pjdtest failure. there is one error in s3.objectprefix, s3.objectPrefix needed.
> 
> yes, s3.objectPrefix is needed, and I have to modify the curveadm's code to support this after this pr is merged just as github can't merge the two repo's pr at the same time.

You should give it an appropriate default value so that it can be created successfully

**Comment by 201341** [comment]
cicheck

**Comment by Cyber-SiKu** [comment]
```bash
create fs success.
fileList: []
E 2023-04-10T17:16:31.497514+0800 987453 disk_cache_base.cpp:135] LoadAllCacheFile, cache read dir is not exist.
E 2023-04-10T17:16:31.497532+0800 987453 disk_cache_read.cpp:123] LoadAllCacheReadFile, load all cacched read file fail ret = -1
E 2023-04-10T17:16:31.497556+0800 987453 disk_cache_manager.cpp:104] load all cache read file error. ret = -1
E 2023-04-10T17:16:31.497562+0800 987453 disk_cache_manager_impl.cpp:47] DiskCacheManagerImpl init error.
E 2023-04-10T17:16:31.497567+0800 987453 client_s3_adaptor.cpp:76] Init disk cache failed
E 2023-04-10T17:16:31.497625+0800 987453 main.cpp:231] init fuse client fail, conf = /curvefs/client/conf/client.conf
Begin to mount fs test to /curvefs/client/mnt/var/lib/jenkins/workspace/curve_failover/pjdtest
```

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cichek

**Comment by 201341** [comment]
cicheck

File: curvefs/conf/tools.conf
Comment (wuhongsong) [inline curvefs/conf/tools.conf:38]: maybe replace 0 with false is better?

File: curvefs/src/client/common/config.h
Comment (wuhongsong) [inline curvefs/src/client/common/config.h:138]: ditto

File: curvefs/proto/common.proto
Comment (wuhongsong) [inline curvefs/proto/common.proto:61]: ditto, bool is better?

File: curvefs/src/client/s3/client_s3_adaptor.cpp
Comment (wuhongsong) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:65]: maybe objectPrefix should be a member of FsInfo just like FSType?

File: curvefs/conf/tools.conf
Comment (201341) [reply curvefs/conf/tools.conf:38]: No, there are 3 kinds of prefixs now, maybe more later. So bool isn't ok. Uint32 or Uint8?

File: curvefs/src/client/s3/client_s3_adaptor.cpp
Comment (201341) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:65]: objectPrefix is inited in the [FuseopInit](https://github.com/opencurve/curve/blob/master/curvefs/src/client/fuse_s3_client.cpp#L63) process  ` ::curvefs::client::common::S3Info2FsS3Option(s3Info, &fsS3Option);` And s3Info just come from FsInfo.

File: curvefs/src/client/s3/disk_cache_base.cpp
Comment (wuhongsong) [inline curvefs/src/client/s3/disk_cache_base.cpp:102]: english is needed

File: curvefs/src/client/s3/disk_cache_read.cpp
Comment (wuhongsong) [inline curvefs/src/client/s3/disk_cache_read.cpp:102]: need return

File: curvefs/conf/tools.conf
Comment (201341) [reply curvefs/conf/tools.conf:38]: ok, I will add some comments here.

Comment (ilixiaocui) [inline curvefs/conf/tools.conf:38]: It may be more appropriate to use `objectPrefixVersion` instead of `objectPrefix`.
Please add a comment to explain the difference between different versions.

In addition, it is recommended to merge 6 commits into 1 for submission
image

Comment (ilixiaocui) [inline curvefs/conf/tools.conf:38]: With this design, users can only use the metadata format of the specified version, and cannot upgrade it?

Comment (201341) [reply curvefs/conf/tools.conf:38]: ok

File: curvefs/src/client/s3/client_s3_adaptor.cpp
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:65]: The acquisition of fsinfo is through https://github.com/opencurve/curve/blob/1f41ffe8c4f601d77f9aab302d5927ed4d8a6902/curvefs/src/client/curve_fuse_op.cpp#L88

The processing on the mds side is in
https://github.com/opencurve/curve/blob/1f41ffe8c4f601d77f9aab302d5927ed4d8a6902/curvefs/src/mds/mds_service.cpp#L199

mds needs to be assigned according to whether the optional prefixVersion exists.

Comment (201341) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:65]: Hi, objectPrefix is inited in the [FuseopInit](https://github.com/opencurve/curve/blob/master/curvefs/src/client/fuse_s3_client.cpp#L63) process  ::curvefs::client::common::S3Info2FsS3Option(s3Info, &fsS3Option); And s3Info just come from FsInfo.

File: curvefs/conf/tools.conf
Comment (201341) [reply curvefs/conf/tools.conf:38]: Yes, users cann't change the prefix once the prefix is decided just as the s3.blocksize. If the users want to change the prefix, There are two methods, one is to create another fs, and sync form the previous fs. The other is to sync the s3 object from one bucket to another bucket, While syncing the object, change the object format. Finally,  using the tools to change the bucket and prefix.

Comment (201341) [reply curvefs/conf/tools.conf:40]: done

File: curvefs/src/client/s3/disk_cache_base.cpp
Comment (201341) [reply curvefs/src/client/s3/disk_cache_base.cpp]: done

File: curvefs/src/client/s3/disk_cache_read.cpp
Comment (201341) [reply curvefs/src/client/s3/disk_cache_read.cpp:102]: done

File: curvefs/conf/tools.conf
Comment (201341) [reply curvefs/conf/tools.conf:40]: I think `objectPrefixVersion ` can't express the meaning. 0 means no prefix, And 1/2/3/... means different kinds of prefix.

**Comment by 201341** [comment]
cicheck

File: curvefs/src/client/s3/client_s3_adaptor.cpp
Comment (ilixiaocui) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:65]: > Hi, objectPrefix is inited in the [FuseopInit](https://github.com/opencurve/curve/blob/master/curvefs/src/client/fuse_s3_client.cpp#L63) process ::curvefs::client::common::S3Info2FsS3Option(s3Info, &fsS3Option); And s3Info just come from FsInfo.

FsInfo of FuseopInit come from 
https://github.com/opencurve/curve/blob/1f41ffe8c4f601d77f9aab302d5927ed4d8a6902/curvefs/src/client/curve_fuse_op.cpp#L88

File: curvefs/conf/tools.conf
Comment (ilixiaocui) [reply curvefs/conf/tools.conf:40]: image

It's not that the prefix is ​​different, but the organizational form of the objectname has changed. Or `objectNameVersion`?

`objectNameVersion=0` means fsid_inodeid_chunkid_index_compaction
`objectNameVersion=1` means fsid_inodeid/1000/1000_inodeid/1000_fsid_inodeid_chunkid_index_compaction

Comment (201341) [reply curvefs/conf/tools.conf:40]: There's no meaning, just adds prefix to improve write and sync performance.

**Review by wuhongsong — APPROVED** [review]
LGTM

File: curvefs/src/client/s3/client_s3_adaptor.cpp
Comment (201341) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:65]: Hi, I think we shouldn't change the mds code, which require upgrading the binary. I can check whether the objectprefix has bee set in  the client, what do you think it is ok ?
`fsS3Opt->objectPrefix = s3.has_objectprefix() ? s3.objectprefix() : 0;`

Comment (ilixiaocui) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:65]: > Hi, I think we shouldn't change the mds code, which require upgrading the binary. I can check whether the objectprefix has bee set in the client, do you think it is ok ? `fsS3Opt->objectPrefix = s3.has_objectprefix() ? s3.objectprefix() : 0;`

LGTM

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by 201341** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Pull request closed without merging**
